playbook/actions/gitops.go

```diff
@@ -0,0 +1,236 @@
+package actions
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/duty/context"
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/gomplate/v3"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/pkg/clients/git"
+	"github.com/flanksource/incident-commander/pkg/clients/git/connectors"
+	gitv5 "github.com/go-git/go-git/v5"
+)
+
+type GitOps struct {
+	workTree *gitv5.Worktree
+	spec     *connectors.GitopsAPISpec
+	env      TemplateEnv
+
+	shouldCreatePR bool
+}
+
+type GitOpsActionResult struct {
+	CreatedPR int `json:"createdPR,omitempty"`
+}
+
+func (t *GitOps) Run(ctx context.Context, action v1.GitOpsAction, env TemplateEnv) (*GitOpsActionResult, error) {
+	var response GitOpsActionResult
+
+	if len(action.Patches) == 0 && len(action.Files) == 0 {
+		logger.Warnf("no patches or files specified on gitops action")
+		return nil, nil
+	}
+
+	t.env = env
+
+	if err := t.generateSpec(ctx, action); err != nil {
+		return nil, err
+	}
+
+	connector, workTree, err := t.cloneRepo(ctx, action)
+	if err != nil {
+		return nil, fmt.Errorf("failed to clone repo: %w", err)
+	}
+	t.workTree = workTree
+
+	if err := t.applyPatches(ctx, action); err != nil {
+		return nil, fmt.Errorf("failed to apply patches: %w", err)
+	}
+
+	if err := t.modifyFiles(ctx, action); err != nil {
+		return nil, fmt.Errorf("failed to modify files: %w", err)
+	}
+
+	if err := git.CommitAndPush(ctx, connector, workTree, t.spec); err != nil {
+		return nil, fmt.Errorf("failed to commit and push: %w", err)
+	}
+
+	if t.shouldCreatePR {
+		prNumber, err := t.createPR(ctx, connector, workTree)
+		if err != nil {
+			return nil, err
+		}
+		response.CreatedPR = prNumber
+	}
+
+	return &response, nil
+}
+
+// generateSpec generates the spec for the git client from the action
+func (t *GitOps) generateSpec(ctx context.Context, action v1.GitOpsAction) error {
+	var err error
+
+	if action.Repo.Base == "" {
+		action.Repo.Base = "master"
+	}
+
+	if action.Repo.Branch == "" {
+		action.Repo.Branch = action.Repo.Base
+	} else {
+		action.Repo.Branch, err = gomplate.RunTemplate(t.env.AsMap(), gomplate.Template{Template: action.Repo.Branch})
+		if err != nil {
+			return err
+		}
+	}
+
+	t.spec = &connectors.GitopsAPISpec{
```

Comment:
Why do we need GitopsAPI Spec ?

Reply:
This came directly from tenant-controller. And I tried to use it as it is an external lib and then convert our spec to `GitopsAPISpec` . But I guess we could just directly use `v1.GitOpsAction`